### PR TITLE
Run2bash

### PR DESCRIPTION
**Pull request by RichWellum**
# US109032 : Vagrant Tooling: CRR: iosxr_setup.py logic additions
- Logic to detect full/mini image and not configure grpc if mini
- Change sudo ALL to vagrant user only
- Configure the resolv.conf in XR (using lokhesh sync code) if possible.
  _resolv.conf logic hasn't been changed. This is because the syncing of /etc/netns/global-vrf/resolv.conf to /etc/resolv.conf requires 'ip netns exec global-vrf bash'. I can't find a reliable way to automate this (tried putting it in .profile - but run into permissions issues). Tried it in .ssh/config but that wasn't possible. So keeping this in xrnns for now._
- Move to using bash instead of run where appropriate
- Fix up ssh inconsistencies - lots of timeouts when coming up.
  _Fixed these by going back to bash -c - as 'bash' causes a shell to be created and I get very odd timing issues. However wrapped these in a new function() which applies the config line and waits for the prompt. Execution of the pexpect code is much clearer and logical now, you can actually follow each line of CLI being executed and getting back a prompt._

Tested full and mini images, k9 and non-k9 images.

Timing is MUCH improved and I don't get any SSH authentication errors anymore.

flake8 satisfied.
# Tests

```
==> default: Machine 'default' has a post `vagrant up` message. This is a message
==> default: from the creator of the Vagrantfile, and not from Vagrant itself:
==> default:
==> default:
==> default:     Welcome to the IOS XRv (64-bit) Virtualbox.
==> default:     To connect to the XR Linux shell, use: 'vagrant ssh'.
==> default:     To ssh to the XR Console, use: 'vagrant port' (vagrant version > 1.8)
==> default:     to determine the port that maps to guestport 22,
==> default:     then: 'ssh vagrant@localhost -p <forwarded port>'
==> default:
==> default:     IMPORTANT:  READ CAREFULLY
==> default:     The Software is subject to and governed by the terms and conditions
==> default:     of the End User License Agreement and the Supplemental End User
==> default:     License Agreement accompanying the product, made available at the
==> default:     time of your order, or posted on the Cisco website at
==> default:     www.cisco.com/go/terms (collectively, the 'Agreement').
==> default:     As set forth more fully in the Agreement, use of the Software is
==> default:     strictly limited to internal use in a non-production environment
==> default:     solely for demonstration and evaluation purposes. Downloading,
==> default:     installing, or using the Software constitutes acceptance of the
==> default:     Agreement, and you are binding yourself and the business entity
==> default:     that you represent to the Agreement. If you do not agree to all
==> default:     of the terms of the Agreement, then Cisco is unwilling to license
==> default:     the Software to you and (a) you may not download, install or use the
==> default:     Software, and (b) you may return the Software as more fully set forth
==> default:     in the Agreement.
rwellum@RWELLUM-M-34DF:[~/Desktop/Boxes]: vagrant ssh
xr-vm_node0_RP0_CPU0:~$ ping google.com
PING google.com (216.58.217.142) 56(84) bytes of data.
64 bytes from iad23s43-in-f142.1e100.net (216.58.217.142): icmp_seq=1 ttl=63 time=9.32 ms
64 bytes from iad23s43-in-f14.1e100.net (216.58.217.142): icmp_seq=2 ttl=63 time=10.1 ms
64 bytes from iad23s43-in-f14.1e100.net (216.58.217.142): icmp_seq=3 ttl=63 time=12.2 ms
64 bytes from iad23s43-in-f14.1e100.net (216.58.217.142): icmp_seq=4 ttl=63 time=9.27 ms
^C
--- google.com ping statistics ---
4 packets transmitted, 4 received, 0% packet loss, time 3006ms
rtt min/avg/max/mdev = 9.274/10.254/12.248/1.210 ms
xr-vm_node0_RP0_CPU0:~$ whoami
vagrant

rwellum@RWELLUM-M-34DF:[~/Desktop/Boxes]: ssh vagrant@localhost -p 2223
The authenticity of host '[localhost]:2223 ([127.0.0.1]:2223)' can't be established.
RSA key fingerprint is b9:3b:13:c7:a5:40:fa:a5:e3:60:5c:8e:dd:5c:61:3a.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added '[localhost]:2223' (RSA) to the list of known hosts.
vagrant@localhost's password:


RP/0/RP0/CPU0:ios#show run
Thu Jun 30 22:43:05.841 UTC
Building configuration...
!! IOS XR Configuration version = 6.1.1.20I
!! Last configuration change at Thu Jun 30 18:32:36 2016 by vagrant
!
telnet vrf default ipv4 server max-servers 10
username vagrant
 group root-lr
 group cisco-support
 secret 5 $1$jgUf$DMwjKda.bsDXbuI7wA0SH/
!
tpa
 address-family ipv4
  update-source MgmtEth0/RP0/CPU0/0
 !
!
interface MgmtEth0/RP0/CPU0/0
 ipv4 address dhcp
!
router static
 address-family ipv4 unicast
  0.0.0.0/0 MgmtEth0/RP0/CPU0/0 10.0.2.2
 !
!
ssh server v2
ssh server vrf default
grpc
 port 57777
!
end
```

### CODE REVIEW

File: iosxr_setup.py [hunk 1]
```diff
@@ -36,6 +36,25 @@ def open_sessions(self, username, password):
         self.xr_telnet_sessions.append(self.xr1)
 
     #
+    # Send a global-vrf XR Linux command - wait for prompt
+    # Could do bash -c or run ip netns exec global-vrf bash
+    #
+    def send_operns(self, command):
+        xr1 = self.xr1
+
+        xr1.send("bash -c %s" % command)
```
Comment (glennmatthews) [inline iosxr_setup.py:44]: Wasn't there some case where you needed to quote the command? `"bash -c '%s'"`?

File: iosxr_setup.py [hunk 2]
```diff
@@ -36,6 +36,25 @@ def open_sessions(self, username, password):
...
+        xr1.wait("[\$#]")
```
Comment (glennmatthews) [inline iosxr_setup.py]: Perhaps:

`return xr1.wait("[\$#]")`

so that you could e.g. below do

`output = send_operns("rpm -qa | grep k9sec")`

File: iosxr_setup.py [hunk 3]
```diff
@@ -36,6 +36,25 @@ def open_sessions(self, username, password):
...
+
+    #
+    # Send a xrnns XR Linux command - wait for prompt
+    #
+    def send_xrnns(self, command):
+        xr1 = self.xr1
+
+        xr1.send("run %s" % command)
+        xr1.wait("[\$#]")
```
Comment (glennmatthews) [inline iosxr_setup.py]: Is the backslash actually needed?

File: iosxr_setup.py [hunk 4]
```diff
@@ -113,60 +146,56 @@ def setup_networking(self, host_ip=None, gateway=None, allow_root_login=False):
...
+        self.send_operns("chkconfig --add sshd_operns")
+
         xr1.wait("RP/0/RP0/CPU0:ios")
 
         # Set up IOS XR ssh if a k9/crypto image
         if k9:
             xr1.send("crypto key generate rsa")
             xr1.wait("How many bits in the modulus")
             xr1.send("")  # Send enter to get default 2048
-            xr1.wait("[\$#]")  # Wait for the prompt
```
Comment (glennmatthews) [inline iosxr_setup.py:169]: Why remove this?

File: iosxr_setup.py (hunk 1 above)
Comment (RichWellum) [inline iosxr_setup.py:44]: I can't recall tbh - lots of changes recently.

File: iosxr_setup.py (hunk 4 above)
Comment (RichWellum) [inline iosxr_setup.py:169]: Noticed this too and put it back in.

File: iosxr_setup.py (hunk 3 above)
Comment (RichWellum) [inline iosxr_setup.py]: Removed and seems to be fine.

File: iosxr_setup.py (hunk 2 above)
Comment (RichWellum) [inline iosxr_setup.py]: Agreed.

File: iosxr_setup.py (hunk 1 above)
Comment (glennmatthews) [inline iosxr_setup.py:44]: Doesn't seem to be needed - faulty memory on my part.

File: iosxr_setup.py [hunk 5]
```diff
@@ -36,6 +36,24 @@ def open_sessions(self, username, password):
...
 
     #
+    # Send a global-vrf XR Linux command - wait for prompt
+    # Could do bash -c or run ip netns exec global-vrf bash
+    #
+    def send_operns(self, command):
+        xr1 = self.xr1
+
+        xr1.send("bash -c %s" % command)
+        time.sleep(1)
```
Comment (glennmatthews) [inline iosxr_setup.py]: shouldn't be needed since you're wait()ing below?

File: iosxr_setup.py [hunk 6]
```diff
@@ -36,6 +36,24 @@ def open_sessions(self, username, password):
...
+        return xr1.wait("[$#]")
+
+    #
+    # Send a xrnns XR Linux command
+    #
+    def send_xrnns(self, command):
+        xr1 = self.xr1
+        xr1.send("run %s" % command)
+
```
Comment (glennmatthews) [inline iosxr_setup.py:53]: Why did you remove the wait() here?

File: iosxr_setup.py [hunk 7]
```diff
@@ -50,10 +68,22 @@ def setup_networking(self, host_ip=None, gateway=None, allow_root_login=False):
...
+        else:
+            xr1.log("Non crypto k9 image detected")
+
+        # Determine if the image is a full or mini by searching for the mgbl rpm
+        # which only exists in the full image
+        # This will be used to configure features only available in a full image
+        output = self.send_operns("rpm -qa | grep mgbl")
+        full = re.search(r'-mgbl', output)
+        if full:
+            xr1.log("Full image detected")
```
Comment (glennmatthews) [inline iosxr_setup.py]: _Technically_ this is checking whether mgbl rpm is installed (which is all we actually need for grpc) not whether it's a full image, so a mini + mgbl image would also meet this requirement - might be good to change "full" to "mgbl" throughout, but non-blocking.

File: iosxr_setup.py [hunk 8]
```diff
@@ -113,60 +141,58 @@ def setup_networking(self, host_ip=None, gateway=None, allow_root_login=False):
...
-        #
-        # Send commands to XR Linux
-        #
-        xr1.send("run")
+            xr1.send("bash -c sed -i 's/PermitRootLogin no/PermitRootLogin yes/' /etc/ssh/sshd_config_operns")
 
         # Add passwordless sudo as required by jenkins
-        xr1.send("echo '####Added by iosxr_setup to give vagrant passwordless access' >> /etc/sudoers")
-        xr1.send("echo '%sudo ALL=(ALL) NOPASSWD: ALL' >> /etc/sudoers")
+        # sudo not vagrant because we are operating in xrnns and global-vrf user space
```
Comment (glennmatthews) [inline iosxr_setup.py:143]: Not sure I understand this comment or the change back from `vagrant` to `%sudo` - can you explain?

File: iosxr_setup.py (hunk 5 above)
Comment (RichWellum) [inline iosxr_setup.py]: Ack.

File: iosxr_setup.py (hunk 7 above)
Comment (RichWellum) [inline iosxr_setup.py]: That's a good change - will do.

File: iosxr_setup.py (hunk 8 above)
Comment (goblinhack) [inline iosxr_setup.py:143]: BTW the echo to sudoers - you mentioned echo issues.

you tried a subshell ?

sudo sh -c "echo \"group ALL=(user) NOPASSWD: ALL\" >> /etc/sudoers"

sunstone.sh does both these approaches
        su -c "cat <<EOF >> /etc/sudoers
$LOGNAME ALL=(ALL:ALL) ALL
$LOGNAME ALL=(ALL) NOPASSWD:ALL
EOF"

```
    echo $LOGNAME ALL=NOPASSWD: ALL | sudo tee -a /etc/sudoers
```

File: iosxr_setup.py (hunk 6 above)
Comment (RichWellum) [inline iosxr_setup.py:53]: Seemed to mess with the resolve.conf.

I'm going to rework this code using sudovi.

File: iosxr_setup.py (hunk 8 above)
Comment (RichWellum) [inline iosxr_setup.py:143]: Seemed to mess with the resolve.conf.

I'm going to rework this code using sudovi.

Comment (RichWellum) [inline iosxr_setup.py:143]: @goblinhack good point - have reworked.

File: iosxr_setup.py [hunk 9]
```diff
@@ -75,15 +74,13 @@ def setup_networking(self, host_ip=None, gateway=None, allow_root_login=False):
         else:
             xr1.log("Non crypto k9 image detected")
 
-        # Determine if the image is a full or mini by searching for the mgbl rpm
-        # which only exists in the full image
-        # This will be used to configure features only available in a full image
+        # Determine if the image has the MGBL package needed for GRPC
```
Comment (glennmatthews) [inline iosxr_setup.py]: nit: "gRPC"

File: iosxr_setup.py [hunk 10]
```diff
@@ -170,9 +167,9 @@ def setup_networking(self, host_ip=None, gateway=None, allow_root_login=False):
         # This will prevent users from needing to supply another Vagrantfile or editing /etc/resolv.conf manually
         # Doing this in xrnns because the syncing of /etc/netns/global-vrf/resolv.conf to
         # /etc/resolv.conf requires 'ip netns exec global-vrf bash'.
-        self.send_xrnns('echo "# Cisco OpenDNS IPv4 nameservers" >> /etc/resolv.conf')
-        self.send_xrnns('echo "nameserver 208.67.222.222" >> /etc/resolv.conf')
-        self.send_xrnns('echo "nameserver 208.67.220.220" >> /etc/resolv.conf')
+        self.send_xrnns("echo '# Cisco OpenDNS IPv4 nameservers' >> /etc/resolv.conf")
```
Comment (glennmatthews) [inline iosxr_setup.py:169]: tee -a here too?

File: iosxr_setup.py (hunk 6 above)
Comment (RichWellum) [inline iosxr_setup.py:53]: Sorry not sudovi - this is used for the resolv.conf.

If I add a delay OR change to using tee -a - then the resolv.conf file fails to update - or is partially updated. I can't explain it. :(

File: iosxr_setup.py (hunk 10 above)
Comment (goblinhack) [inline iosxr_setup.py:169]: tee -a might be better in general - it gets around all these permission
worries of doing
stuff in 2 steps...

File: iosxr_setup.py (hunk 9 above)
Comment (RichWellum) [inline iosxr_setup.py]: Done.

File: iosxr_setup.py (hunk 10 above)
Comment (RichWellum) [inline iosxr_setup.py:169]: Something about doing this in xrnns - very touchy. Leaving as is for now.

File: iosxr_setup.py (hunk 1 above)
Comment (RichWellum) [inline iosxr_setup.py:44]: @glennmatthews @goblinhack I think this is the only part that I haven't changed - xrnns and resolv.conf. I'll keep looking at it but let me know if other changes look ok?

**Comment by glennmatthews** [comment]
LGTM
